### PR TITLE
Add tracing support for gin

### PR DESCRIPTION
**Pull request by ringsaturn**
Fix https://github.com/getsentry/sentry-go/discussions/593

### CODE REVIEW

File: gin/sentrygin.go [hunk 1]
```diff
@@ -12,12 +13,65 @@ import (
...
-const valuesKey = "sentry"
+const (
+	valuesKey      = "sentry"
+	traceValuesKey = "sentry-gin"
+
+	// https://develop.sentry.dev/sdk/performance/#header-sentry-trace
+	traceHeader = "sentry-trace"
+
+	// https://develop.sentry.dev/sdk/performance/dynamic-sampling-context/#baggage
+	baggageHeader = "baggage"
```
Comment (cleptric) [inline gin/sentrygin.go:24]: We should move this into a central place.

File: gin/sentrygin.go [hunk 2]
```diff
@@ -30,6 +84,18 @@ type Options struct {
 	WaitForDelivery bool
 	// Timeout for the event delivery requests.
 	Timeout time.Duration
+	// Extract Sentry trace id from request, by default use `sentry-trace` from header
+	GetTraceIDFromRequest func(*gin.Context) string
```
Comment (cleptric) [inline gin/sentrygin.go:88]: This shouldn't be an option.

File: gin/sentrygin.go [hunk 3]
```diff
@@ -30,6 +84,18 @@ type Options struct {
...
+	// Extract Sentry baggage from request, by default use `baggage` from header
+	GetBaggageFromRequest func(*gin.Context) string
```
Comment (cleptric) [inline gin/sentrygin.go:90]: This shouldn't be an option.

File: gin/sentrygin.go [hunk 4]
```diff
@@ -52,9 +126,21 @@ func (h *handler) handle(ctx *gin.Context) {
 		hub = sentry.CurrentHub().Clone()
 	}
 	hub.Scope().SetRequest(ctx.Request)
+
+	transaction := sentry.StartTransaction(
+		ctx, fmt.Sprintf("%v %v", ctx.Request.Method, ctx.FullPath()),
+		sentry.ContinueFromHeaders(h.getTraceIDFromRequest(ctx), h.getBaggageFromRequest(ctx)),
+	)
+	ctx.Writer.Header().Set(traceHeader, transaction.ToSentryTrace())
+	ctx.Writer.Header().Set(baggageHeader, transaction.ToBaggage())
```
Comment (cleptric) [inline gin/sentrygin.go:135]: Can you explain why we need to do this?

File: gin/sentrygin.go [hunk 5]
```diff
@@ -52,9 +126,21 @@ func (h *handler) handle(ctx *gin.Context) {
 		hub = sentry.CurrentHub().Clone()
 	}
 	hub.Scope().SetRequest(ctx.Request)
+
+	transaction := sentry.StartTransaction(
+		ctx, fmt.Sprintf("%v %v", ctx.Request.Method, ctx.FullPath()),
+		sentry.ContinueFromHeaders(h.getTraceIDFromRequest(ctx), h.getBaggageFromRequest(ctx)),
+	)
```
Comment (cleptric) [inline gin/sentrygin.go:133]: We'll also need to set a source, `sentry.TransctionSource(sentry.SourceURL),`.
Please construct the options above and pass them in.

File: gin/sentrygin.go [hunk 6]
```diff
@@ -12,12 +13,65 @@ import (
...
+
+	// https://develop.sentry.dev/sdk/performance/dynamic-sampling-context/#baggage
+	baggageHeader = "baggage"
+)
+
+// Convert HTTP Status to [Sentry status], rewrite from Sentry Python SDK's [tracing.py]
+//
+// [Sentry status]: https://develop.sentry.dev/sdk/event-payloads/properties/status/
+// [tracing.py]: https://github.com/getsentry/sentry-python/blob/1.12.0/sentry_sdk/tracing.py#L436-L467
+func FromHTTPStatusToSentryStatus(code int) sentry.SpanStatus {
```
Comment (cleptric) [inline gin/sentrygin.go:31]: Might be worth moving this into a central place.

File: gin/sentrygin.go [hunk 7]
```diff
@@ -96,3 +182,12 @@ func GetHubFromContext(ctx *gin.Context) *sentry.Hub {
 	}
 	return nil
 }
+
+// StartSpanFromGinContext start a new *sentry.Span from gin.Context.
+func StartSpanFromGinContext(ctx *gin.Context, op string) *sentry.Span {
```
Comment (cleptric) [inline gin/sentrygin.go:187]: We shouldn't expose more ways to start a span. If `StartSpan` can't handle this use-case, we should fix it.

File: gin/sentrygin.go (hunk 4 above)
Comment (ringsaturn) [reply gin/sentrygin.go]: I’m not sure whether it’s useful to return trace info in HTTP header, like Sentry’ Python SDK works with Flask&FastAPI.

Is there any spec for this use case?

File: gin/sentrygin.go (hunk 7 above)
Comment (ringsaturn) [reply gin/sentrygin.go:124]: For now, in `sentrygin` package, there is no way to access this struct:

https://github.com/getsentry/sentry-go/blob/27e50ad6c8d6c892e2f968ca21061431793bb1ee/tracing.go#L841-L842

Any idea?

File: gin/sentrygin.go [hunk 8]
```diff
@@ -52,9 +56,31 @@ func (h *handler) handle(ctx *gin.Context) {
...
+			ctx.GetHeader(sentry.SentryBaggageHeader),
+		),
+		sentry.TransctionSource(sentry.SourceURL),
+	}
+	transaction := sentry.StartTransaction(
+		ctx, fmt.Sprintf("%v %v", ctx.Request.Method, ctx.FullPath()),
+		options...,
+	)
+
+	ctx.Request = ctx.Request.WithContext(transaction.Context())
```
Comment (ringsaturn) [inline gin/sentrygin.go:72]: @cleptric I add a print in `StartSpan` func:

```diff
diff --git a/tracing.go b/tracing.go
index 96df72b..1f8fbbe 100644
--- a/tracing.go
+++ b/tracing.go
@@ -97,6 +97,7 @@ type TraceParentContext struct {
 // transaction to Sentry.
 func StartSpan(ctx context.Context, operation string, options ...SpanOption) *Span {
    parent, hasParent := ctx.Value(spanContextKey{}).(*Span)
+   fmt.Printf("hasParent = %v\n", hasParent)
    var span Span
    span = Span{
        // defaults
```

Got false all the time: 

```
hasParent = false
```

**Comment by cleptric** [comment]
Closing in favour of https://github.com/getsentry/sentry-go/pull/644

**Pull request closed without merging**